File: scripts/repro_scenario16/probe.cpp
// Standalone replay of the C++ ratelimit final update, from exact operand bits.
#include <cstdio>
#include <cstdint>
#include <cstring>
#include <algorithm>
#include <cstdlib>
static double D(const char*h){uint64_t u=strtoull(h,nullptr,16);double d;memcpy(&d,&u,8);return d;}
static uint64_t B(double d){uint64_t u;memcpy(&u,&d,8);return u;}
int main(int c,char**v){
  double inp=D(v[1]),L=D(v[2]),DT=D(v[3]),minR=D(v[4]),maxR=D(v[5]);
  double rate=(inp-L)/DT;
  rate=std::min(std::max(rate,minR),maxR);      // saturate_c
  double res=L+rate*DT;
  printf("C++     rate=0x%016llX  result=0x%016llX\n",
         (unsigned long long)B(rate),(unsigned long long)B(res));
  return 0;
}

File: scripts/repro_scenario16/probe.f90
! Standalone replay of ROSCO's ratelimit final update, from exact operand bits.
! Args (hex Z16): inputSignal LastSignal DT minRate maxRate
program probe
  implicit none
  integer, parameter :: DbKi = 8
  character(len=32) :: a(5)
  integer(8) :: z(5), zr, zo
  real(DbKi) :: inp, L, DT, minR, maxR, rate, res
  integer :: i
  do i=1,5
    call get_command_argument(i, a(i)); read(a(i),'(Z16)') z(i)
  end do
  inp=transfer(z(1),inp); L=transfer(z(2),L); DT=transfer(z(3),DT)
  minR=transfer(z(4),minR); maxR=transfer(z(5),maxR)
  rate = (inp - L)/DT
  rate = real(min(max(rate,minR),maxR), DbKi)   ! ROSCO saturate()
  res  = L + rate*DT
  zr=transfer(rate,zr); zo=transfer(res,zo)
  write(*,'(A,Z16.16,A,Z16.16)') 'FORTRAN rate=0x',zr,'  result=0x',zo
end program
